sw/device/lib/dif/dif_keymgr.h

```diff
@@ -0,0 +1,648 @@
+ * boot stage. A typical usage of this library during different secure boot
+ * stages is as follows:
+ *
+ * - In Mask ROM:
+ *   - Create a new handle: `dif_keymgr_init()`.
+ *   - Initialize state: `dif_keymgr_initialize_state()`,
+ *   `dif_keymr_get_status()`.
+ *   - Advance state: `dif_keymgr_advance_state()`, `dif_keymgr_get_status()`.
+ * - In subsequent boot stages, i.e. ROM_EXT, BL0, kernel:
+ *   - Create a new handle: `dif_keymgr_init()`.
+ *   - Generate keys and/or identity seeds:
+ *     `dif_keymgr_generate_versioned_key()`,
+ *     `dif_keymgr_generate_identity_seed()`, `dif_keymgr_get_status()`.
+ *   - Read output: `dif_keymgr_read_output()`.
+ *   - Advance state: `dif_keymgr_advance_state()`, `dif_keymgr_get_status()`.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param[out] keymgr Out-param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_init(dif_keymgr_params_t params,
+                                    dif_keymgr_t *keymgr);
+
+/**
+ * Initializes key manager after power-on reset.
+ *
+ * After PoR, key manager is in Reset state with a non-deterministic secret
+ * value. This function causes key manager to initialize its secret value using
+ * the random values generated by the entropy source and transition to
+ * Initialized state. Note that Initialized state is not an operational
+ * state and key manager state must be advanced one more time by calling
+ * `dif_keymgr_advance_state()` before keys or identity seeds can be generated.
+ *
+ * This is an asynchronous function because key manager state transitions
+ * involve KMAC operations that can take some time to complete. Clients must
+ * check the status of key manager using `dif_keymgr_get_status()` before
+ * calling other functions in this library.
+ *
+ * The entropy source must be initialized before this function is called because
+ * key manager relies on the entropy source to initialize its secret value.
+ *
+ * This function must be called only once per power cycle.
+ *
+ * @param keymgr A key manager handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_initialize_state(const dif_keymgr_t *keymgr);
+
+/**
+ * Parameters for advancing the state of key manager.
+ */
+typedef struct dir_keymgr_advance_state_params {
+  /**
+   * This value is used by key manager to derive secret values and can be either
+   * a value that represents the contents of a boot stage, e.g. a truncated
+   * hash, or a tag. If it is a hash, changes in a boot stage will change the
+   * secret value, and consequently the versioned keys and identity seeds
+   * generated at subsequent boot stages. If it is a tag, those secret values,
+   * versioned keys, and identity seeds will be preserved across updates of the
+   * boot stage as long as the tag remains the same.
+   */
+  uint32_t sw_binding[DIF_KEYMGR_SW_BINDING_WORDS];
```

Comment:
Is there a particular reason this isn't a byte array? I assume the alignment isn't tremendously important.